src/librustc/middle/stability.rs

```diff
@@ -663,7 +711,40 @@ fn lookup_uncached<'tcx>(tcx: &ty::ctxt<'tcx>, id: DefId) -> Option<&'tcx Stabil
                 // unmarked impls for it. See FIXME above for more details.
 
                 debug!("lookup: trait_id={:?}", trait_id);
-                return lookup(tcx, trait_id);
+                return lookup_stability(tcx, trait_id);
+            }
+        }
+        None
+    })
+}
+
+fn lookup_deprecation_uncached<'tcx>(tcx: &ty::ctxt<'tcx>, id: DefId) -> Option<Deprecation> {
+    debug!("lookup(id={:?})", id);
+
+    // is this definition the implementation of a trait method?
+    match tcx.trait_item_of_item(id) {
+        Some(ty::MethodTraitItemId(trait_method_id)) if trait_method_id != id => {
+            debug!("lookup: trait_method_id={:?}", trait_method_id);
+            return lookup_deprecation(tcx, trait_method_id)
+        }
+        _ => {}
+    }
+
+    let item_depr = if id.is_local() {
+        None // The stability cache is filled partially lazily
+    } else {
+        tcx.sess.cstore.deprecation(id)
+    };
+
+    item_depr.or_else(|| {
+        if tcx.is_impl(id) {
+            if let Some(trait_id) = tcx.trait_id_of_impl(id) {
+                // FIXME (#18969): for the time being, simply use the
+                // stability of the trait to determine the stability of any
+                // unmarked impls for it. See FIXME above for more details.
+
+                debug!("lookup: trait_id={:?}", trait_id);
+                return lookup_deprecation(tcx, trait_id);
```

Comment:
I've removed some obsolete code from these functions instead, should be better now.